rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,225 @@
+process and don't need to maintain a new special image and EKS installation job.
+
+### Automatic Discovery and Enrollment
+
+Users will also be able to set up automatic discovery and enrollment of EKS clusters in the Discover UI.
+We will reuse the existing EKS discovery capability, although a new mode will be added to it - discovery through
+the AWS integration. EKS discovery in that mode will rely on the AWS integration, selected in the Discovery UI,
+to find clusters and then install the Helm chart on them. The algorithm will be as follows:
+- List EKS clusters through the AWS integration.
+- Cross-check to see which EKS clusters are already registered in the system, using labels we inject on Helm chart installation.
+- Once a new EKS cluster is found, associate our OIDC provider with it and wait for association to be in active state.
+- Using the associated provider, generate an access token and install the kube agent through the Helm chart.
+- Disassociate our OIDC provider from the EKS cluster.
+
+Users will be able to filter EKS clusters for discovery by using labels.
+
+A dynamic discovery configuration will be created as a result of users setting up automatic EKS discovery in the Discover UI (see [RFD 125](https://github.com/gravitational/teleport/blob/master/rfd/0125-dynamic-auto-discovery-config.md)).
+This configuration can be picked up by any discovery service. In the Teleport cloud, the Discovery service will be available by default,
+meaning users will not need to perform any additional actions.
+
+### EKS Identity Provider Association
+
+AWS EKS supports a method of authentication that involves associating an OIDC identity provider with the cluster.
+Such an associated identity provider can issue access tokens (JWT) that grant access to the EKS cluster groups and users.
+Since Teleport already supports OIDC providers, we can easily use this approach. It will require slight
+modifications to the AWS OIDC integration code to allow it to issue JWTs for an EKS cluster. To associate an OIDC provider with an
+EKS cluster, the selected AWS integration needs to issue an API call `eks:AssociateIdentityProviderConfig`
+with the following configuration (includes example parameters):
+
+```yaml
+---
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+
+metadata:
+  name: eks-cluster-name
+  region: us-east-1
+
+identityProviders:
+  - name: TeleportProvider
+    type: oidc
+    issuerUrl: https://teleport.example.com
+    clientId: kubernetes
+    usernameClaim: teleport-eks-oidc-user
+    groupsClaim: teleport-eks-oidc-groups
+```
+
+To access the target EKS cluster in API mode, we will generate an access token using the associated OIDC provider.
+That token will grant access to the `system:masters` group, so the Helm chart installation will be able to set up
+all the required infrastructure for the Teleport kube agent, including service accounts.
+
+AWS supports the association of only one OIDC identity provider with an EKS
```

Comment:
Would we be confident enough to temporarily replace the OIDC until the chart is installed and in the end we swap back to the old configuration? Maybe it's too intrusive

Teleport doesn't need to have long access to the cluster so we should anw remove our configured access to it once we install helm chart.

Reply:
I thought about it, but associating OIDC provider itself takes 3-4 minutes. So overall we would break access to their existing clients (that rely on old OIDC for auth) for ~10 minutes until old ODIC is restored. I think it's way too much.